### ISSUE TITLE
Backup helper script fails silently when name resolution fails

### ISSUE THREAD

**Issue by ppmathis** [mailcow/mailcow-dockerized#3913]
**Prior to placing the issue, please check following:** *(fill out each checkbox with an `X` once done)*
- [X] I understand that not following or deleting the below instructions will result in immediate closure and/or deletion of my issue.
- [X] I have understood that this bug report is dedicated for bugs, and not for support-related inquiries.
- [X] I have understood that answers are voluntary and community-driven, and not commercial support.
- [X] I have verified that my issue has not been already answered in the past. I also checked previous [issues](https://github.com/mailcow/mailcow-dockerized/issues).

## Summary
The backup helper script (`helper-scripts/backup_and_restore.sh`) might silently fail when attempting to backup the MySQL database due to an IPv6 implementation detail of Docker. I've attempted to setup our backup solution to automatically call the helper script before running the actual remote backup job and noticed that the MySQL backup step sporadically fails.

Upon closer inspection it turns out to be a sporadic name resolution failure happening when `mariabackup` attempts to connect to the MySQL database server `mysql`. While this sometimes works, the script often fails with the log messages listed below. While I was first clueless what might cause this issue, attaching `nicolaka/netshoot` and running tcpdump to the container running `mariabackup` revealed what was going wrong.

When the Docker host has IPv4 and IPv6 nameservers configured in its `/etc/resolv.conf` (in the example below I've used Cloudflares 1.1.1.1 service) **AND** uses `options rotate`, Docker ends up generating a `resolv.conf` like this for its containers:

```
nameserver 127.0.0.11
nameserver 2606:4700:4700::1111
nameserver 2606:4700:4700::1001
options [...] rotate [...]
```

As can be seen here, the original IPv4 resolvers of the host system (`1.1.1.1` + `1.0.0.1`) were replaced with the internal IPv4 resolver of Docker, whereas the IPv6 resolvers have been passed through as-is. This sometimes results in an attempted resolution of `mysql.` against public IPv6 nameservers (due to `rotate` forcing the system to cycle resolvers), resulting in `NXDOMAIN` and ultimately bubbling up the chain causing `mariabackup` to fail.

I was not able to find any solution for this issue and unfortunately Docker does not feature an embedded IPv6 DNS server as can be seen in https://github.com/moby/moby/issues/41651, which also documents related failures. I did however find three workarounds:

1) Changing the backup helper script to run `while ! mariabackup ...; do sleep 1; done` inside the container, causing it to retry the backup several times - sometimes it works on first run, sometimes it takes up to 10 failed attempts. This is obviously just a dirty solution and would require an actual upper limit of retries, so I am not using it.

2) Remove all IPv6 nameservers from the host system, resulting in only `127.0.0.11` being listed as a resolver within the container, guaranteeing the resolution of `mysql` to work as long as the containers are all up and running. As I do want to use dual-stacked resolvers though, I am not using this either.

3) Get rid of `options rotate` within the host resolver configuration. This guarantees that `127.0.0.11` will always be queried first because Docker orders IPv4 nameservers before IPv6 nameservers, ignoring the host-preferred order. As the impact is near-zero, this was my go-to solution and works like a charm.

While I am not sure if there is any easy fix for Mailcow itself as it seems related to upstream / how Docker handles name resolution, I would at least expect the backup helper script to fail with a non-zero exit code. Unfortunately as of today this is not the case and a database backup can silently fail due to this issue.

### Backup helper script
```
Using /opt/mailcow-backups as backup/restore location.

Found project name mailcow
Using SQL image mariadb:10.4, starting...
[00] 2020-12-25 19:28:16 Connecting to MySQL server host: mysql, user: root, password: set, port: 3306, socket: /var/run/mysqld/mysqld.sock
[00] 2020-12-25 19:28:16 Failed to connect to MySQL server: Unknown MySQL server host 'mysql' (-2).
mariabackup based on MariaDB server 10.4.17-MariaDB debian-linux-gnu (x86_64)
mariabackup: Can't change dir to '/backup_mariadb/' (errno: 2 "No such file or directory")
[00] 2020-12-25 19:28:16 can't my_setwd /backup_mariadb/
chown: cannot access '/backup_mariadb': No such file or directory
/bin/tar: /backup_mariadb: Cannot stat: No such file or directory
/bin/tar: Exiting with failure status due to previous errors
```

### Traffic capture of working run
Taken using `docker run -it --rm --network container:$(docker ps | grep mailcow-backup | awk '{print $1}') nicolaka/netshoot tcpdump -niany port 53`:
```
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
19:45:11.591864 IP 127.0.0.11.53 > 127.0.0.1.58740: 44974 1/0/0 AAAA fd4d:6169:6c63:6f77::c (56)
19:45:11.591916 IP 127.0.0.11.53 > 127.0.0.1.58740: 33105 1/0/0 A 172.22.1.9 (44)
```

### Traffic capture of failed run
Taken using `docker run -it --rm --network container:$(docker ps | grep mailcow-backup | awk '{print $1}') nicolaka/netshoot tcpdump -niany port 53`:
```
19:46:16.808403 IP6 fd4d:6169:6c63:6f77::12.35399 > 2606:4700:4700::1001.53: 47022+ A? mysql. (23)
19:46:16.808650 IP6 fd4d:6169:6c63:6f77::12.35399 > 2606:4700:4700::1001.53: 52138+ AAAA? mysql. (23)
19:46:16.829786 IP6 2606:4700:4700::1001.53 > fd4d:6169:6c63:6f77::12.35399: 52138 NXDomain 0/1/0 (98)
19:46:16.829827 IP6 2606:4700:4700::1001.53 > fd4d:6169:6c63:6f77::12.35399: 47022 NXDomain 0/1/0 (98)
19:46:16.830529 IP 172.22.1.14.38626 > 1.1.1.1.53: 8622+ AAAA? mysql.<system domain>. (42)
19:46:16.830532 IP 172.22.1.14.60429 > 1.1.1.1.53: 11178+ A? mysql.<system domain>. (42)
19:46:16.852661 IP 1.1.1.1.53 > 172.22.1.14.38626: 8622 NXDomain 0/1/0 (109)
19:46:16.852930 IP 172.22.1.14.54996 > 1.0.0.1.53: 8622+ AAAA? mysql.<system domain>. (42)
19:46:16.853532 IP 1.1.1.1.53 > 172.22.1.14.60429: 11178 NXDomain 0/1/0 (109)
19:46:16.853759 IP 172.22.1.14.47048 > 1.0.0.1.53: 11178+ A? mysql.<system domain>. (42)
19:46:16.874930 IP 1.0.0.1.53 > 172.22.1.14.54996: 8622 NXDomain 0/1/0 (109)
19:46:16.875196 IP 127.0.0.11.53 > 127.0.0.1.54106: 8622 NXDomain 0/1/0 (135)
19:46:16.875524 IP 1.0.0.1.53 > 172.22.1.14.47048: 11178 NXDomain 0/1/0 (109)
19:46:16.875635 IP 127.0.0.11.53 > 127.0.0.1.54106: 11178 NXDomain 0/1/0 (135)
```

## Reproduction
I've been able to reliable reproduce this on a system running Debian Buster 10 with the latest Docker release taken from the official repositories, which is `20.10.1` as of today. The easiest way to reproduce this is by adding dual-stacked resolvers to the host with `options rotate`, followed by running `MAILCOW_BACKUP_LOCATION=<whatever> /opt/mailcow-dockerized/helper-scripts/backup_and_restore.sh backup mysql` several times which results in sporadic failures.

For running tcpdump against the container (as described above) adding a delay, e.g. `sleep 2` before the `mariabackup` command can be helpful. To ensure it is not related to my specific environment, I also reproduced this on a fresh Debian 10 VM on Hetzner Cloud:

1) Deploy CX31 and add IPv4/IPv6 resolvers (e.g. Cloudflare) while using `options rotate`:
```shell
cat >/etc/resolv.conf <<EOF
nameserver 1.1.1.1
nameserver 1.0.0.1
nameserver 2606:4700:4700::1111
nameserver 2606:4700:4700::1001
options rotate
EOF
```

2) Install Mailcow with Docker and Docker Compose according to docs: https://mailcow.github.io/mailcow-dockerized-docs/i_u_m_install/

3) Wait for Mailcow containers to fully initialize

4) Run the backup command stated at the beginning of the reproduction section several times with a valid backup folder, which results in a sporadic failure.

## Expectations
- The backup script should propagate the failure by exiting with a non-zero status code.
- The docs might profit from a hint to avoid `options rotate` in the host resolver configuration, as this might lead to sporadic name resolution failures like this which are hard to debug.

## System information

| Question | Answer |
| --- | --- |
| My operating system | Debian Buster x86-64 |
| Is Apparmor, SELinux or similar active? | AppArmor is active |
| Virtualization technlogy (KVM, VMware, Xen, etc - **LXC and OpenVZ are not supported** | KVM |
| Server/VM specifications (Memory, CPU Cores) | 8GB of RAM, 2 CPU cores |
| Docker Version (`docker version`) | 20.10.1 |
| Docker-Compose Version (`docker-compose version`) | 1.27.4 |
| Reverse proxy (custom solution) | nginx |

- Output of `git diff origin/master`, any other changes to the code? If so, **please post them**.
No changes have been made.
- All third-party firewalls and custom iptables rules are unsupported. *Please check the Docker docs about how to use Docker with your own ruleset*. Nevertheless, iptabels output can help us to help you: `iptables -L -vn`, `ip6tables -L -vn`, `iptables -L -vn -t nat` and `ip6tables -L -vn -t nat`.
No custom rules or firewalls are being used, fully managed by Docker.
- DNS problems? Please run `docker exec -it $(docker ps -qf name=acme-mailcow) dig +short stackoverflow.com @172.22.1.254` (set the IP accordingly, if you changed the internal mailcow network) and post the output.
Not related but for the sake of completeness:
```
151.101.65.69
151.101.129.69
151.101.193.69
151.101.1.69
```

**Comment by andryyy** [comment]
That's interesting, thanks.

**Comment by andryyy** [comment]
Can you try it with the workaround? I have not found a way to remove or override a DNS option as given by the host. Perhaps disabling IPv6 for the backup will work.

**Comment by ppmathis** [comment]
I've enabled `options rotate` on my system again, successfully reproduced the initially reported issue with the old/previous version and then updated to the newest version. I double-checked that the changes of 46928269bbdf7c113b517a8fe2c65c7be4d01aee are present in my deployment and was no longer to reproduce this issue.

I ran a total of about 25 DB backups and not a single one failed with the issue which previously happened frequently, so it seems like this workaround does the job. Thank you for your work and happy new year!
